### ISSUE TITLE
Binary string (`BYTEA`, `Binary`) concatenation

### ISSUE THREAD

**Issue by kawadakk** [apache/datafusion#12709]
### Is your feature request related to a problem or challenge?

Currently there doesn't seem to be a direct way to concatenate binary strings in SQL expressions. 

- `lhs || rhs` does not accept `Binary` arguments.
- `concat(lhs, rhs)` coerces both sides to `Utf8`, which does not work if they are not valid UTF-8 strings.
- `array_concat(lhs, rhs)` only works for `List`, not `Binary`.
- `decode(encode(lhs, 'hex') || encode(rhs, 'hex'), 'hex')` works, but this is verbose and inefficient.

### Describe the solution you'd like

The `||` operator could be extended to support `Binary` arguments, following the behavior of other SQL engines, such as SQLite and [PostgreSQL](https://www.postgresql.org/docs/current/functions-binarystring.html).

**Comment by alamb** [comment]
I agree this still seems not to be supported

```sql
> select arrow_typeof(x'1234');
+-------------------------------+
| arrow_typeof(Binary("18,52")) |
+-------------------------------+
| Binary                        |
+-------------------------------+
1 row(s) fetched.
Elapsed 0.003 seconds.

> select x'1234' || x'5678';
Error during planning: Cannot infer common string type for string concat operation Binary || Binary
```

**Comment by jayzhan211** [comment]
ArrayConcat use signature `variadic_any` which should be changed to something in ArraySignature

**Comment by Jefffrey** [comment]
Some pointers:

Look here for the existing `StringConcat` operator:

https://github.com/apache/datafusion/blob/a02e6836e7d449008eef592b177ea282595e027a/datafusion/expr-common/src/operator.rs#L83-L84

We should look at places it is used and either try to extend it to include binary (although it might be weird since the name is `StringConcat` unless we rename it) or potentially introduce a new operator `BinaryConcat`. We should also consider the UDF:

https://github.com/apache/datafusion/blob/a02e6836e7d449008eef592b177ea282595e027a/datafusion/functions/src/string/concat.rs

One thing to keep in mind is how we handle `binary || string` concat (i.e. which common type we coerce to); we can't assume all binary is a valid string, but strings are always valid binary.

**Comment by devanshu0987** [comment]
> One thing to keep in mind is how we handle binary || string concat (i.e. which common type we coerce to); we can't assume all binary is a valid string, but strings are always valid binary.

```
devanshu=# SELECT 'abc'::text || '\x00'::bytea AS result,  pg_typeof('abc'::text || '\x00'::bytea) AS concat_result;
 result  | concat_result 
---------+---------------
 abc\x00 | text
(1 row)
```

I validated the rules in Postgres 18. It coerces it to Text type regardless of whether the binary string is valid or not.
bytea is coerced to text via its textual representation.

Can we follow the same for Data Fusion?

**Comment by Jefffrey** [comment]
I don't think we can follow that behaviour for DataFusion, because our version of casting binary to string will try to interpret the actual bytes as UTF8, whereas it seems postgres will convert it to the hex representation?

**Comment by devanshu0987** [comment]
> it seems postgres will convert it to the hex representation

Yes. I used Cursor to find the implementation. Seems it does a hex_encode.
`src/backend/utils/adt/bytea.c`

```
Datum
byteaout(PG_FUNCTION_ARGS)
{
	bytea	   *vlena = PG_GETARG_BYTEA_PP(0);
	char	   *result;
	char	   *rp;

	if (bytea_output == BYTEA_OUTPUT_HEX)
	{
		/* Print hex format */
		rp = result = palloc(VARSIZE_ANY_EXHDR(vlena) * 2 + 2 + 1);
		*rp++ = '\\';
		*rp++ = 'x';
		rp += hex_encode(VARDATA_ANY(vlena), VARSIZE_ANY_EXHDR(vlena), rp);
	}
```

**Comment by Jefffrey** [comment]
I think we can proceed with assumption of preferring binary over string, and maybe with a PR up there can be discussion from others on how it looks and if it's preferred to instead cast to string in cases of `binary || string` 🤔

**Comment by theirix** [comment]
It's safer to make a proper binary concatenation for a binary/string mix. Spark does it more safely this way, while the majority (Postgres, DuckDB, and Snowflake) actually prefer casting to text for historical reasons.

I can draft a PR with an experimental implementation for both binary/string and binary/binary to verify that it fits our expectations.

**Comment by alamb** [comment]
In my opinion we should follow what the other systems do rather than inventing our own  semantics

**Comment by theirix** [comment]
What do you think of having a dialect-dependent implementation via customising `ExprPlanner`? For Spark/Databricks we plan a binary | binary -> text, while for others, we use binary. So we can have expected behaviour for every platform.

**Comment by Jefffrey** [comment]
> What do you think of having a dialect-dependent implementation via customising `ExprPlanner`? For Spark/Databricks we plan a binary | binary -> text, while for others, we use binary. So we can have expected behaviour for every platform.

Do we have existing precedent for such customization at that level?

**Comment by theirix** [comment]
> > What do you think of having a dialect-dependent implementation via customising `ExprPlanner`? For Spark/Databricks we plan a binary | binary -> text, while for others, we use binary. So we can have expected behaviour for every platform.
> 
> Do we have existing precedent for such customization at that level?

Not quiet. The idea is to make a plan using two UDFs in `ExprPlanner::plan_binary_op` - one `ConcatBinary` with text-to-binary and the classical `Concat`. The planner chooses between them based on a dialect. A good example is `NestedFunctionPlanner::plan_binary_op`

Since it had not been done before, I stumbled upon maintaining an actual config with a dialect in the expression planner, so we can react to `set dialect` dynamically. A session state cannot be passed to `SessionStateBuilder::with_default_features`, where default planners are instantiated early, while the state is rebuilt often via `new_from_existing`. I believe it could be done with some redesign, but it's worth confirming if it's worth it.

**Comment by Jefffrey** [comment]
Maybe it's best to keep this simple and follow what Postgres does for now; we do have precedent for having Spark specific functions so if we want to follow Spark we could have that behaviour in the datafusion-spark crate 🤔

**Comment by theirix** [comment]
@Jefffrey I implemented a Postgres-like behaviour in this PR. If it looks good, I can also follow-up with a opt-in Spark binary-to-binary implementation

**Comment by theirix** [comment]
take

**Comment by theirix** [comment]
Resolution: went with casting binaries to text - the most common behaviour except Spark

**Comment by kawadakk** [comment]
@theirix Thank you for the fix, but `Binary || Binary` still evaluates to `Utf8`, which does not work if they are not valid UTF-8 strings and does not match PostgreSQL's behavior:

```console
$ datafusion-cli
> select arrow_typeof(x'31' || x'32');
+--------------------------------------------+
| arrow_typeof(Binary("49") || Binary("50")) |
+--------------------------------------------+
| Utf8                                       |
+--------------------------------------------+
1 row(s) fetched.
Elapsed 0.003 seconds.

> select x'ff' || x'af';
Optimizer rule 'simplify_expressions' failed
caused by
Arrow error: Invalid argument error: Encountered non UTF-8 data: invalid utf-8 sequence of 1 bytes from index 0
```

```console
$ psql

postgres=# select decode('31', 'hex') || decode('32', 'hex');
 ?column?
----------
 \x3132
(1 row)

postgres=# select decode('ff', 'hex') || decode('af', 'hex');
 ?column?
----------
 \xffaf
(1 row)
```

**Comment by theirix** [comment]
I suggest reopening this issue to discuss semantics again.

The issue is that the logic was introduced for the N-ary `concat` UDF, but not to the binary pipe operator (`||`) - they return text but have entirely different code.  While `concat` allows invalid UTF-8 in binary inputs by deferring validation to the entire concatenated binary, the pipe operator doesn't do so. The optimiser tries to optimise this plan
```
Projection: CAST(Binary("255") AS Utf8) || CAST(Binary("175") AS Utf8)
      EmptyRelation: rows=1
```
and [fails trying to evaluate literals](https://github.com/apache/datafusion/blob/206ac6b950e709373a39e74ebb33ed0c95edce9e/datafusion/optimizer/src/simplify_expressions/expr_simplifier.rs#L550). The same error with `concat`. It doesn't happen when arguments are valid UTF-8 binaries as per added SLTs - https://github.com/apache/datafusion/compare/main...theirix:datafusion:binary-concat-fixes . Of course, if we follow Postgres' behaviour with hexifying binary output, it won't happen.

While I was summarising what other engines do, I discovered that Postgres is actually an outlier, and it handles binaries by emitting a hex representation, as pointed out above. This Postgres implementation for both UDF and operator emits binary as decoded text, not as hexified string `SELECT x'c3a9' || 'hello'::text` -> `éhello` instead of `1100001110101001hello` with Postgres.

Engine | str+str | str+bin | bin+bin
-- | -- | -- | --
PostgreSQL | text | Mixed text and hexified bytes | Binary (if both args are bytea) or hexified str
ClickHouse | String of bytes | String of bytes | String of bytes
Spark/Databricks | String | Type error (must cast) | Binary
DuckDB | String | Type error (must cast) | Binary
Snowflake | String | Type error (must cast) | Binary
MySQL | String | Binary string | Binary string

Now I'm wondering if we should provide bin+bin -> bin logic by changing `concat` UDF and providing a totally new `BinaryConcat` operator instead of `StringConcat` when both operands are binaries. It depends on the choice between Postgres and various AnalyticalDBs.

**Comment by theirix** [comment]
> I suggest reopening this issue to discuss semantics again.
> 
> The issue is that the logic was introduced for the N-ary `concat` UDF, but not to the binary pipe operator (`||`) - they return text but have entirely different code. While `concat` allows invalid UTF-8 in binary inputs by deferring validation to the entire concatenated binary, the pipe operator doesn't do so. The optimiser tries to optimise this plan
> 
> ```
> Projection: CAST(Binary("255") AS Utf8) || CAST(Binary("175") AS Utf8)
>       EmptyRelation: rows=1
> ```
> 
> and [fails trying to evaluate literals](https://github.com/apache/datafusion/blob/206ac6b950e709373a39e74ebb33ed0c95edce9e/datafusion/optimizer/src/simplify_expressions/expr_simplifier.rs#L550). The same error with `concat`. It doesn't happen when arguments are valid UTF-8 binaries as per added SLTs - [main...theirix:datafusion:binary-concat-fixes](https://github.com/apache/datafusion/compare/main...theirix:datafusion:binary-concat-fixes) . Of course, if we follow Postgres' behaviour with hexifying binary output, it won't happen.
> 
> While I was summarising what other engines do, I discovered that Postgres is actually an outlier, and it handles binaries by emitting a hex representation, as pointed out above. This Postgres implementation for both UDF and operator emits binary as decoded text, not as hexified string `SELECT x'c3a9' || 'hello'::text` -> `éhello` instead of `1100001110101001hello` with Postgres.
> Engine 	str+str 	str+bin 	bin+bin
> PostgreSQL 	text 	Mixed text and hexified bytes 	Binary (if both args are bytea) or hexified str
> ClickHouse 	String of bytes 	String of bytes 	String of bytes
> Spark/Databricks 	String 	Type error (must cast) 	Binary
> DuckDB 	String 	Type error (must cast) 	Binary
> Snowflake 	String 	Type error (must cast) 	Binary
> MySQL 	String 	Binary string 	Binary string
> 
> Now I'm wondering if we should provide bin+bin -> bin logic by changing `concat` UDF and providing a totally new `BinaryConcat` operator instead of `StringConcat` when both operands are binaries. It depends on the choice between Postgres and various AnalyticalDBs.

@Jefffrey @alamb @jayzhan211 I would like to ask for your opinion on these findings and on which path we should follow.

**Comment by Jefffrey** [comment]
Maybe it's better to only support str+str=str and bin+bin=bin for concat & `||`, and any mixed ones (str+bin, bin+str) should be a type error like spark/duckdb/snowflake. That seems like the least surprising behaviour to me.

**Comment by theirix** [comment]
Binary operator support has been merged, and I am working on a final part to synchronise the operator and UDF behaviour.

**Comment by theirix** [comment]
This feature is finally done. All concat UDFs (generic concat, concat_ws and Spark concat) and the pipe operator are fully consistent:
- we allow binary vs binary concatenation, producing a binary type.
- if the arguments are strings, it works as before, producing a string.
- if the arguments are mixed types (strings vs binary), then binaries are coerced to strings.

It matches the behaviour of other analytical databases and Postgres - all of them actually cast binaries to text and produce a text output for mixed cases (correcting previous findings in the table above https://github.com/apache/datafusion/issues/12709#issuecomment-4194430343)


@kawadakk, regarding your latest comment, I believe the behaviour is expected now
```
DataFusion CLI v54.0.0
+--------------------------------------------+
| arrow_typeof(Binary("49") || Binary("50")) |
+--------------------------------------------+
| Binary                                     |
+--------------------------------------------+
1 row(s) fetched.
Elapsed 0.025 seconds.

+------------------------------+
| Binary("49") || Binary("50") |
+------------------------------+
| 3132                         |
+------------------------------+
1 row(s) fetched.
Elapsed 0.007 seconds.

+--------------------------------+
| Binary("255") || Binary("175") |
+--------------------------------+
| ffaf                           |
+--------------------------------+
1 row(s) fetched.
Elapsed 0.006 seconds.
```
